In [1]:
from IPython.display import HTML, display

def bgc(color, cell=None):
    script = (
        "var cell = this.closest('.jp-CodeCell');"
        "var editor = cell.querySelector('.jp-Editor');"
        "editor.style.background='{}';"
        "this.parentNode.removeChild(this)"
    ).format(color)

    display(HTML('<img src onerror="{}">'.format(script)))



<div class="alert alert-block alert-danger"><font color=black>

> Carl: I was going to avoid the color coding issue on "Day One" but upon rereading, maybe it's get it over with.
    
> I've modified the Your Turns below to be color coded.
    
> I also put in an explanation about how to complete the Your Turns and the Homework.  Perhaps there is better language for this? 
    
</font>  
</div>

This is a Jupyter lab notebook.

It has two types of cells:

* Python cells, and 
* Markdown cells.

Python cells are preceded by `[]:` possibly with a number in the brackets.

Markdown cells are not preceded by anything.  They look different depending on whether they have been compiled or not.

This is a markdown cell. It has probably been compiled when you opened this notebook.  

> Double click on this cell to see what the markdown code looks like.

Execute this cell by pressing the $\blacktriangleright$ symbol at the top of this tab.

The keyboard shortcut to execute is `shift` + `enter(return)`.

There is some markdown code in the cell above. 

The  \\$\blacktriangleright\\$  piece is latex.

You don't need to worry about this code for now.  You'll naturally pick it up as you find you want to use it.

You *will* need to use markdown cells to provide answers to questions *and* provide interpretation of output results.  Writing and running code is *never* enough - you *always* must communicate what you've done and what the output tells you.

When you insert a new cell, it will default to a Python cell.  Toggle back and forth between Python (Code) cells and Mardown cells using the dropdown at the top of this tab.

#### Your turn:

> Turn the empty Python cell below into a Markdown cell.

> Write a sentence or two describing your experience with computer languages and coding. 

> Good communication is important; write a grammatically correct, complete sentence.

#### Answer:

<div class="alert alert-block alert-info"><font color=black>

I did some html writing when I was in college, and also learned LaTex and some MatLab.
    
</font>  
</div>

#### Python:

Next, we'll experiment a little with Python code. 

A few words of advice first:

* The notebooks in this course contain Python code.  You *must* think carefully about the code provided as examples.  Do **not** simply execute the code and move on.
* The examples are just that - examples.  They are not meant to be able to solve all problems.  Rather, you should think of the commands you learn in the examples as tools to use, in different combinations, to solve problems.
* Python is **huge**!  We are not going to try to learn *all* of Python upfront.  Instead, we will learn Python on as-needed basis to solve our Data Science needs.
* The solutions presented as examples in the notebooks are by no means the only (or even the best) way to solve the problems we present.  You are encouraged to find other/better solutions.

So, let's see what Python looks like and get a general idea how it works.  As promised above, we'll jump right in with some data that we'll analyze further in the next notebook.

To analyze data in Python, you have to load it into this notebook.  The command `read_csv` will do this for us.  But, `read_csv` is part of a package called `pandas`.  

Below, we'll first `install` pandas. You only need to do this once in a notebook.  Then we'll load the `.csv` file.

In [2]:
import pandas as pd

In [3]:
pd.read_csv('https://raw.githubusercontent.com/CarlLienert/DS-FLC/main/can_langs.csv')

Unnamed: 0                                 category  \
0             0                     Aboriginal languages   
1             1  Non-Official & Non-Aboriginal languages   
2             2  Non-Official & Non-Aboriginal languages   
3             3  Non-Official & Non-Aboriginal languages   
4             4  Non-Official & Non-Aboriginal languages   
..          ...                                      ...   
209         209  Non-Official & Non-Aboriginal languages   
210         210                     Aboriginal languages   
211         211  Non-Official & Non-Aboriginal languages   
212         212  Non-Official & Non-Aboriginal languages   
213         213  Non-Official & Non-Aboriginal languages   

                           language  mother_tongue  most_at_home  \
0      Aboriginal languages, n.o.s.            590           235   
1                         Afrikaans          10260          4785   
2    Afro-Asiatic languages, n.i.e.           1150           445   
3                        Akan (Twi)          13460          5985   
4                          Albanian          26895         13135   
..                              ...            ...           ...   
209                           Wolof           3990          1385   
210                      Woods Cree           1840           800   
211               Wu (Shanghainese)          12915          7650   
212                         Yiddish          13555          7085   
213                          Yoruba           9080          2615   

     most_at_work  lang_known  
0              30         665  
1              85       23415  
2              10        2775  
3              25       22150  
4             345       31930  
..            ...         ...  
209            10        8240  
210            75        2665  
211           105       16530  
212           895       20985  
213            15       22415  

[214 rows x 7 columns]

Since `read_csv` is part of the `pandas` package, we have to preface the `read_csv` command with `pd.` 

(`.pd` is a standard abbreviation for `pandas` used to shorten the code a little)

The part of the `read_csv` command in red is the address of the `.csv` file.  We'll talk more about that issue later.

Go back and look at the `read_csv` command.  It loaded *and* displayed output for us to see the contents of the`.csv` file.  However, if we can't conveniently *do* anything with this output, if seeing the contents of the file is the end goal, great - we're done. If we want to do something with the file, then we should give this output a name, so we can reference it:

In [4]:
df_can_langs=pd.read_csv('https://raw.githubusercontent.com/CarlLienert/DS-FLC/main/can_langs.csv')

This command line produced no output because we didn't ask it to.  It's a good idea to just type the name of the file to make sure the command did the *right* thing. 

In [5]:
df_can_langs

Unnamed: 0                                 category  \
0             0                     Aboriginal languages   
1             1  Non-Official & Non-Aboriginal languages   
2             2  Non-Official & Non-Aboriginal languages   
3             3  Non-Official & Non-Aboriginal languages   
4             4  Non-Official & Non-Aboriginal languages   
..          ...                                      ...   
209         209  Non-Official & Non-Aboriginal languages   
210         210                     Aboriginal languages   
211         211  Non-Official & Non-Aboriginal languages   
212         212  Non-Official & Non-Aboriginal languages   
213         213  Non-Official & Non-Aboriginal languages   

                           language  mother_tongue  most_at_home  \
0      Aboriginal languages, n.o.s.            590           235   
1                         Afrikaans          10260          4785   
2    Afro-Asiatic languages, n.i.e.           1150           445   
3                        Akan (Twi)          13460          5985   
4                          Albanian          26895         13135   
..                              ...            ...           ...   
209                           Wolof           3990          1385   
210                      Woods Cree           1840           800   
211               Wu (Shanghainese)          12915          7650   
212                         Yiddish          13555          7085   
213                          Yoruba           9080          2615   

     most_at_work  lang_known  
0              30         665  
1              85       23415  
2              10        2775  
3              25       22150  
4             345       31930  
..            ...         ...  
209            10        8240  
210            75        2665  
211           105       16530  
212           895       20985  
213            15       22415  

[214 rows x 7 columns]

A few comments about the name:
* `df` is for *data frame* - that's what this file is.
* try to pick names that give an idea what they represent (Canadian languages in this case) 
* try to pick names that are not too long
* don't reuse the same name

Notice the data frame contains 214 rows and 7 columns, but the default output does not show all the rows.  Say we want to see the information in the 63rd row:

In [6]:
df_can_langs.iloc[63]

Unnamed: 0                                            63
category         Non-Official & Non-Aboriginal languages
language                                           Ganda
mother_tongue                                       1295
most_at_home                                         345
most_at_work                                          25
lang_known                                          2495
Name: 63, dtype: object

Actions we want to take on data frames often have this syntax.  We start with the name of the data frame, `df_can_langs`, and append the *action* to the end with a `.`

I'm pretty sure true Python programmers would cringe at this explanation.  A more accurate description is something like: `df_can_langs` is an *object* and `iloc` is one of the object's *attributes*.  In this course, we'll be more practical than philosophical.

You might have noticed something interesting in the *actions* we have performed.

* When we uploaded the `csv` file the syntax was `action(*argument goes here*)`
* When we looked for the 63rd row the syntax was `object.*action goes here*`

How do you know which one to use?  The only way I know which one to use is by experience and by trying.  

`df_can_langs` is a Data Frame, in particular, a *pandas* Data Frame.  Pandas provides some powerful, and dangerous, ways to filter the data.  Execute each of the following cells and compare the output.

In [7]:
column1=df_can_langs[['language']]
column1

language
0      Aboriginal languages, n.o.s.
1                         Afrikaans
2    Afro-Asiatic languages, n.i.e.
3                        Akan (Twi)
4                          Albanian
..                              ...
209                           Wolof
210                      Woods Cree
211               Wu (Shanghainese)
212                         Yiddish
213                          Yoruba

[214 rows x 1 columns]

In [8]:
column2=df_can_langs['language']
column2

0        Aboriginal languages, n.o.s.
1                           Afrikaans
2      Afro-Asiatic languages, n.i.e.
3                          Akan (Twi)
4                            Albanian
                    ...              
209                             Wolof
210                        Woods Cree
211                 Wu (Shanghainese)
212                           Yiddish
213                            Yoruba
Name: language, Length: 214, dtype: object

In [9]:
column3=df_can_langs['language'].values.tolist()
column3

['Aboriginal languages, n.o.s.',
 'Afrikaans',
 'Afro-Asiatic languages, n.i.e.',
 'Akan (Twi)',
 'Albanian',
 'Algonquian languages, n.i.e.',
 'Algonquin',
 'American Sign Language',
 'Amharic',
 'Arabic',
 'Armenian',
 'Assyrian Neo-Aramaic',
 'Athabaskan languages, n.i.e.',
 'Atikamekw',
 'Austro-Asiatic languages, n.i.e',
 'Austronesian languages, n.i.e.',
 'Azerbaijani',
 "Babine (Wetsuwet'en)",
 'Bamanankan',
 'Beaver',
 'Belarusan',
 'Bengali',
 'Berber languages, n.i.e.',
 'Bikol',
 'Bilen',
 'Blackfoot',
 'Bosnian',
 'Bulgarian',
 'Burmese',
 'Cantonese',
 'Carrier',
 'Catalan',
 'Cayuga',
 'Cebuano',
 'Celtic languages, n.i.e.',
 'Chaldean Neo-Aramaic',
 'Chilcotin',
 'Chinese languages, n.i.e.',
 'Chinese, n.o.s.',
 'Comox',
 'Cree, n.o.s.',
 'Creole languages, n.i.e.',
 'Creole, n.o.s.',
 'Croatian',
 'Cushitic languages, n.i.e.',
 'Czech',
 'Dakota',
 'Danish',
 'Dene',
 'Dinka',
 'Dogrib (Tlicho)',
 'Dravidian languages, n.i.e.',
 'Dutch',
 'Edo',
 'English',
 'Estonian',

All three convey the same information.  However, the information is in a different format in each case.  Use the following commands to see what that format is:

In [10]:
type(column1)

pandas.core.frame.DataFrame

In [11]:
type(column2)

pandas.core.series.Series

In [12]:
type(column3)

list

If our goal is to see the values in the `language` column, any of the three is fine. What we want to do next determines which format we want.  We'll deal with that issue as it arises; for now we *generally* want the output to be a Data Frame.  So, in this case, the first of the three commands above: `column1` is a data frame.

#### Your turn:
> Produce a data frame that contains *only* the two columns `mother_tongue` and `lang_known`.

> A note on the cells where you contribute: These notebooks will contain empty cells for your contribution.  Some coding is included to color the cells - different coding for python cells and markdown cells.  Please leave the coding in the cells.

> Python cells will begin with the command `bgc('lightblue')`.  Put your code *after* this command line.

> Markdown cells will have some html code at the beginning and the end. Put your answer *between* the provided html code.

In [13]:
bgc('lightblue')
twocolumns=df_can_langs[['mother_tongue', 'lang_known']]
twocolumns

mother_tongue  lang_known
0              590         665
1            10260       23415
2             1150        2775
3            13460       22150
4            26895       31930
..             ...         ...
209           3990        8240
210           1840        2665
211          12915       16530
212          13555       20985
213           9080       22415

[214 rows x 2 columns]

#### Your turn:

Execute the code below and explain in the markdown box that follows what the command did.

In [14]:
df_can_langs.loc[[3,4],['language', 'mother_tongue', 'lang_known']]

language  mother_tongue  lang_known
3  Akan (Twi)          13460       22150
4    Albanian          26895       31930

#### Answer:

<div class="alert alert-block alert-info"><font color=black>

    
The commmand returned the 3rd and 4th rows, and the three columns `language`, `mother_tongue`, and `lang_known`.
    
</font>  
</div>

Above we've specified rows in an arbitrary manner, there was no particular reason to look at the 3rd and 4th rows.  This is probably not very useful in practice.  

A practical situation might be to restrict our attention to those languages known by a minimum number of people, say 9025.  We're looking for those rows where the number in the `lang_known` column is at least 9025.  The syntax is similar to specifying columns in that the criteria goes inside the square brackets:


In [15]:
df_can_langs[df_can_langs['lang_known']>=9025]

Unnamed: 0                                 category  \
1             1  Non-Official & Non-Aboriginal languages   
3             3  Non-Official & Non-Aboriginal languages   
4             4  Non-Official & Non-Aboriginal languages   
7             7  Non-Official & Non-Aboriginal languages   
8             8  Non-Official & Non-Aboriginal languages   
..          ...                                      ...   
201         201  Non-Official & Non-Aboriginal languages   
204         204  Non-Official & Non-Aboriginal languages   
211         211  Non-Official & Non-Aboriginal languages   
212         212  Non-Official & Non-Aboriginal languages   
213         213  Non-Official & Non-Aboriginal languages   

                   language  mother_tongue  most_at_home  most_at_work  \
1                 Afrikaans          10260          4785            85   
3                Akan (Twi)          13460          5985            25   
4                  Albanian          26895         13135           345   
7    American Sign Language           2685          3020          1145   
8                   Amharic          22465         12785           200   
..                      ...            ...           ...           ...   
201                    Urdu         210815        128785          1495   
204              Vietnamese         156430        104245          8075   
211       Wu (Shanghainese)          12915          7650           105   
212                 Yiddish          13555          7085           895   
213                  Yoruba           9080          2615            15   

     lang_known  
1         23415  
3         22150  
4         31930  
7         21930  
8         33670  
..          ...  
201      322220  
204      198895  
211       16530  
212       20985  
213       22415  

[86 rows x 7 columns]

This command requires a bit more thought:

* The part inside brackets: `df_can_langs['lang_known']>=9025` is called a *boolean*.  That is, it is either True or False.  If it is True, the corresponding row is part of the output. If it is False, the corresponding row is **not** part of the output.
* `df_can_langs['lang_known']` specifies the column of interest.
* `>=9025` gives the criteria on the values of interest in that column

#### Your turn:

Execute the following command. Do you get the same output?  If not, why not?  (Answer goes in markdown cell.)

In [16]:
df_can_langs[df_can_langs['lang_known']>9025]

Unnamed: 0                                 category  \
1             1  Non-Official & Non-Aboriginal languages   
3             3  Non-Official & Non-Aboriginal languages   
4             4  Non-Official & Non-Aboriginal languages   
7             7  Non-Official & Non-Aboriginal languages   
8             8  Non-Official & Non-Aboriginal languages   
..          ...                                      ...   
201         201  Non-Official & Non-Aboriginal languages   
204         204  Non-Official & Non-Aboriginal languages   
211         211  Non-Official & Non-Aboriginal languages   
212         212  Non-Official & Non-Aboriginal languages   
213         213  Non-Official & Non-Aboriginal languages   

                   language  mother_tongue  most_at_home  most_at_work  \
1                 Afrikaans          10260          4785            85   
3                Akan (Twi)          13460          5985            25   
4                  Albanian          26895         13135           345   
7    American Sign Language           2685          3020          1145   
8                   Amharic          22465         12785           200   
..                      ...            ...           ...           ...   
201                    Urdu         210815        128785          1495   
204              Vietnamese         156430        104245          8075   
211       Wu (Shanghainese)          12915          7650           105   
212                 Yiddish          13555          7085           895   
213                  Yoruba           9080          2615            15   

     lang_known  
1         23415  
3         22150  
4         31930  
7         21930  
8         33670  
..          ...  
201      322220  
204      198895  
211       16530  
212       20985  
213       22415  

[85 rows x 7 columns]

#### Answer:

<div class="alert alert-block alert-info"><font color=black>

No, the output is not the same. The first filter has one more row than the second.  This is because there is one language with 9025 speakers.
    
</font>  
</div>

#### Your turn:

Filter the data frame to the row with the language 'Dene'. 

Hint: if your command doesn't work at first, read the error message and consult with a neighbor. 

In [17]:
bgc('lightblue')
df_can_langs[df_can_langs['language']=='Dene']

Unnamed: 0              category language  mother_tongue  most_at_home  \
48          48  Aboriginal languages     Dene          10700          7710   

    most_at_work  lang_known  
48           770       13060

#### Answer:

<div class="alert alert-block alert-danger"><font color=black>

> Carl's: everything below is "red".
    
</font>  
</div>

#### Homework:

> The following data is borrowed from *Introduction to Modern Statistics* (https://openintro-ims.netlify.app/index.html)

> The data set contains information about *notable* earthquakes during the 20th century. (It does not contain information about *all* earthquakes of the 20th century.)

> Load the data frame.  The address is given, but modify the cell so that the data frame has a name.

In [18]:
bgc('lightblue')
df_earthquake=pd.read_csv('https://raw.githubusercontent.com/CarlLienert/DS-FLC/main/earthquakes.csv')
df_earthquake

Year      Month  Day  Richter                         Area  \
0    1902      April   19      7.5  Quezaltenango and San Marco   
1    1902   December   16      6.4                   Uzbekistan   
2    1903      April   28      7.0                    Malazgirt   
3    1903        May   28      5.8                         Gole   
4    1905      April    4      7.5                       Kangra   
..    ...        ...  ...      ...                          ...   
118  1998        May   30      6.6                       Border   
119  1998       July   17      7.0                        Papua   
120  1999    January   25      6.1                      Armenia   
121  1999     August   17      7.6                        Izmit   
122  1999  September   20      7.6                     Taichung   

                     Region   Deaths  
0                 Guatemala   2000.0  
1                    Russia   4700.0  
2                    Turkey   3500.0  
3                    Turkey   1000.0  
4                     India  19000.0  
..                      ...      ...  
118  Afghanistan-Tajikistan   4000.0  
119              New Guinea   2183.0  
120                Colombia   1185.0  
121                  Turkey  17118.0  
122                  Taiwan   2400.0  

[123 rows x 7 columns]

> Create a new data frame that contains only the columns `Region` and `Deaths`.

In [19]:
bgc('lightblue')
earthquake_deaths=df_earthquake[['Region','Deaths']]
earthquake_deaths

Region   Deaths
0                 Guatemala   2000.0
1                    Russia   4700.0
2                    Turkey   3500.0
3                    Turkey   1000.0
4                     India  19000.0
..                      ...      ...
118  Afghanistan-Tajikistan   4000.0
119              New Guinea   2183.0
120                Colombia   1185.0
121                  Turkey  17118.0
122                  Taiwan   2400.0

[123 rows x 2 columns]

> Filter the data frame you created above to show only those notable earthquakes that resulted in 1,000,000 or more deaths.  How many such earthquakes were there? What general region of the world has suffered the most?

In [20]:
bgc('lightblue')
earthquake_deaths[earthquake_deaths['Deaths']>=100000]

Region    Deaths
17         China  200000.0
20         Japan  142800.0
50  Turkmenistan  110000.0
72          Peru  700000.0
84         China  255000.0

#### Answer:

<div class="alert alert-block alert-info"><font color=black>

There were 5 notable earthquakes that resulted in 1,000,000 or more deaths. Four of these earthquakes occured in Asia. Peru, however, suffered by far the most deadly earthquake.
    
</font>  
</div>

> From your answer to the question above, which country has had the most deadly earthquake?  In what year did this earthquake occur?  Hint: which data frame will you use to answer the second question? (You will need to write some code, read the output, and then answer the question.)

In [21]:
bgc('lightblue')
df_earthquake[df_earthquake['Region']=='Peru']

Year     Month  Day  Richter      Area Region    Deaths
47  1946  November   10      7.3    Ancash   Peru    1400.0
72  1970       May   31      7.9  Chimbote   Peru  700000.0

#### Answer:

<div class="alert alert-block alert-info"><font color=black>

The earthquake in Peru that resulted in 7,000,000 deaths occured in 1970.
    
</font>  
</div>

> Has Chile or Peru had more earthquakes? Think about how to filter the data frame and then answer the question.

In [22]:
bgc('lightblue')
earthquake_deaths[earthquake_deaths['Region']=='Peru']

Region    Deaths
47   Peru    1400.0
72   Peru  700000.0

In [23]:
bgc('lightblue')
earthquake_deaths[earthquake_deaths['Region']=='Chile']

Region   Deaths
8   Chile   3882.0
36  Chile  28000.0
61  Chile   1655.0

#### Answer:

<div class="alert alert-block alert-info"><font color=black>

    
Chile has had 3 earthquakes.  Chile has had more earthquakes than Peru, but Peru more deaths.
    
</font>  
</div>